src/lib.rs

```diff
@@ -74,51 +75,115 @@ impl Notifier {
-        (
-            account.to_str().unwrap().to_string(),
-            container.to_str().unwrap().to_string(),
-            key.to_str().unwrap().to_string(),
-            host.to_str().unwrap().to_string()
-        )
+    if let Some(v) = map.remove("azurite_host") {
+        let mut azurite_host = url::Url::parse(&v)
+            .map_err(|e| anyhow!("failed to parse azurite_host: {}", e))?;
+        azurite_host.set_path("");
+        std::env::set_var("AZURITE_BLOB_STORAGE_URL", azurite_host.as_str());
+        map.insert("allow_invalid_certificates".into(), "true".into());
+        map.insert("azure_storage_use_emulator".into(), "true".into());
     }
+
+    if let Some(v) = map.remove("minio_host") {
+        let mut minio_host = url::Url::parse(&v)
+            .map_err(|e| anyhow!("failed to parse minio_host: {}", e))?;
+        minio_host.set_path("");
+        // std::env::set_var("AZURITE_BLOB_STORAGE_URL", azurite_host.as_str());
+        map.insert("allow_http".into(), "true".into());
+        map.insert("aws_endpoint".into(), minio_host.as_str().trim_end_matches('/').to_string());
+        tracing::warn!("aws endpoint: {}", map["aws_endpoint"]);
+    }
+
+    let mut retry_config = RetryConfig::default();
+    if let Some(value) = map.remove("max_retries") {
+        retry_config.max_retries = value.parse()
+            .map_err(|e| anyhow!("failed to parse max_retries: {}", e))?;
+    }
+    if let Some(value) = map.remove("retry_timeout_secs") {
+        retry_config.retry_timeout = std::time::Duration::from_secs(value.parse()
+            .map_err(|e| anyhow!("failed to parse retry_timeout_sec: {}", e))?
+        );
+    }
+
+    let client: Arc<dyn ObjectStore> = match url.scheme() {
+        "s3" => {
+            let mut builder = object_store::aws::AmazonS3Builder::default()
+                .with_url(url)
+                .with_retry(retry_config);
+            for (key, value) in map {
+                builder = builder.with_config(key.parse()?, value);
+            }
+            Arc::new(builder.build()?)
+        }
+        "az" | "azure" => {
+            let mut builder = object_store::azure::MicrosoftAzureBuilder::default()
+                .with_url(url)
+                .with_retry(retry_config);
+            for (key, value) in map {
+                builder = builder.with_config(key.parse()?, value);
+            }
+            Arc::new(builder.build()?)
+        }
+        _ => unimplemented!("unknown url scheme")
+    };
+    // let (store, _) = object_store::parse_url_opts(&url, pairs)?;
+
+    // let client: Arc<dyn ObjectStore> = Arc::new(store);
+
+    // let ping_path: Path = "_this_file_does_not_exist".try_into().unwrap();
+    // match client.get(&ping_path).await {
+    //     Ok(_) | Err(object_store::Error::NotFound { .. }) => {},
+    //     Err(e) => {
+    //         return Err(anyhow!("failed to check store client connection: {}", e));
+    //     }
+    // }
```

Comment:
don't forget to clean this up before merge

Reply:
I left it there so we can discuss this a little bit.

The purpose of the ping here was to ensure the connection was valid before we add it to the cache. Without it we don't have a correctness issue, but we may end up with a miss-configured client living in the `Cache` for a while.

Thinking a little more about it, this is not a big problem. We can enable TTL for the entries of the  `Cache` later if we have too many bad clients.

Reply:
thinking about the cache a little

perhaps the cache size is something we might want to be in StaticConfig?

Also, just to check my understanding (and we might want to document this?): 
- there is a single global cache
- the cache is a [moka::future::Cache](https://docs.rs/moka/latest/moka/future/struct.Cache.html), which is some version of a LRU cache(?) that "will try to evict entries that have not been used recently or very often" when at capacity
- clients are re-used only if the `config_string` hashes to the same value... i.e. the `config_string` needs to be identical for a client to be reused (we don't e.g. iterate the key/values and hash them in a deterministic order, we just hash the whole string exactly as it was passed to us)
- currently, clients are only removed from the cache when the capacity is reached; they do not expire
- currently, the cache has a fixed max capacity of 10

Reply:
Your understanding is correct! Some notes:
- Yes, cache_size should live in StaticConfig. I will add it.
- We use moka as it is a feature complete concurrent cache that supports async.
- Yes, we rely on having a deterministic order of the entries in config_string. But this is only for efficiency, it does not impact correctness. The intended usage pattern is to build the Config struct on the Julia side once and reuse it multiple times, without serializing the config again on every call, so the order will always be the same. A problem can only happen if the Config struct is being recreated every time and only if the Dict implementation has unstable ordering semantics (I know it does not preserve insertion order, but with the same entries the order should remain the same). If we deem necessary one way to prevent this would be to sort the entries and use an `OrderedDict` to serialize them to JSON on the Julia side. Doing something here would incur the extra runtime cost of parsing the JSON and sorting the entries on every call.
- We can prevent old clients from living forever by configuring the cache with a TTL (from StaticConfig)